### DISCUSSION TITLE
StateNode entry event run 2 times when use invoke~

### DISCUSSION THREAD

**Discussion by michael829** [statelyai/xstate#3219] [General]
Hi, I'm test for invoke with child machine, find that sub machine's initial node's entry event auto called. And when main machine's initial state set to the invoke, I receive 3 events(main_entry, child_entry, child_entry).

I'm test to change main machine's initial state to other stateNode, and run send() action to set current to the invoke, but still get 3 events(main_entry, child_entry, child_entry).

here is the code i have run test, please tell me how to get only 2 event when use invoke with child machine? I think child machine's initial event is't Unnecessary with invoke but i can't distinguish and pass it(or neglect it).

`
import { assign, createMachine, interpret } from 'xstate';

const child = createMachine({
  id: 'child',
  initial: 'first',
  context: {
    count: 0,
  },
  states: {
    first: {
      entry: 'onChildEntry'
    }
  }
}, {
  actions: {
    onChildEntry:assign({
      count: (ctx) => {
        console.log('child event');
        return 202;
      }
    })
  }
})

const main = createMachine({
  id: 'main',
  initial: 'child',
  context: {
    count: 0,
  },
  states: {
    child: {
      id: 'my-child',
      invoke: child,
      entry: 'onMainEntry'
    },
  }
}, {
  actions: {
    onMainEntry: assign({
      count: (ctx)=>{
        console.log('main event');
        return 101;
      }
    })
  }
});

const service = interpret(main).start();
// main event
// child event
// child event
`

**Comment by davidkpiano** [comment]
✏️ Please create a [CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed) reproduction of this. Thanks!

**Reply by michael829** [reply]
sorry, i made a sample code in the reply below.

**Comment by michael829** [comment]
Oh sorry...here is a CodeSandbox sample.
[Event order test with invoke!](https://codesandbox.io/s/nostalgic-liskov-p4ovqi?file=/src/index.js)

So, please help me why "main event" fired after "child event"? and how can i reverse this? 
thanks...

**Reply by Andarist** [reply]
You can't easily do that - you would have to somehow restructure your machine to achieve that. The question is - why does it matter for your use case?

**Reply by michael829** [reply]
oh.. I think that event in Xstate strictly limited to onion model. I have write something with Xstate. When i'm start splitting a huge state machine into some sub state machines, the event order not fire as expected with invoke. Maybe I shouldn't rely too much event order? but i still think that node's entry event must fire before invoke's initial events...isn't it?

**Reply by Andarist** [reply]
> Maybe I shouldn't rely too much event order? 

I would say that this is probably true - if you isolate machines then the event order usually shouldn't matter and if you need to coordinate stuff then it should be done explicitly. 

> but i still think that node's entry event must fire before invoke's initial events...isn't it?

Are you talking about the same situation as depicted on the codesandbox or about something else?

**Reply by michael829** [reply]
> before

well.. Oh, I see.. Thank you for your reply

Although I always think that whether the machine is isolated or not, it should always trigger the entry event first and then the initial event of the machine, which seems to be unimportant. (intuitively, this is really, um... It's very difficult to understand why the entry event of the invoke node is after the initial event of the child state machine? It seems that the child state machine is not a child node?) ok I think we need to consider other ways to avoid it

However, shouldn't the events of the sub node of the normal event model be triggered after the parent node Isn't that more correct? I think the sequence of this event is more important

Uh Maybe my idea is wrong ha-ha...

**Reply by Andarist** [reply]
I guess it would be best if you could present a real-world situation in which this makes a difference for you. It's hard to discuss this in an abstract - because in an abstract it just doesn't matter. Fundamentally it could be one way or another - this order is sort of an implementation detail and only observable because JS is single-threaded
